hibernate-core/src/main/java/org/hibernate/event/jfr/internal/JfrEventManager.java

```diff
@@ -0,0 +1,432 @@
+			if ( jdbcConnectionReleaseEvent.shouldCommit() ) {
+				jdbcConnectionReleaseEvent.executionTime = getExecutionTime( jdbcConnectionReleaseEvent.startedAt );
+				jdbcConnectionReleaseEvent.sessionIdentifier = getSessionIdentifier( session );
+				jdbcConnectionReleaseEvent.tenantIdentifier = tenantId;
+				jdbcConnectionReleaseEvent.commit();
+			}
+		}
+	}
+
+	public static JdbcPreparedStatementCreationEvent beginJdbcPreparedStatementCreationEvent() {
+		final JdbcPreparedStatementCreationEvent jdbcPreparedStatementCreation = new JdbcPreparedStatementCreationEvent();
+		if ( jdbcPreparedStatementCreation.isEnabled() ) {
+			jdbcPreparedStatementCreation.begin();
+			jdbcPreparedStatementCreation.startedAt = System.nanoTime();
+		}
+		return jdbcPreparedStatementCreation;
+	}
+
+	public static void completeJdbcPreparedStatementCreationEvent(
+			JdbcPreparedStatementCreationEvent jdbcPreparedStatementCreation,
+			String preparedStatementSql) {
+		if ( jdbcPreparedStatementCreation.isEnabled() ) {
+			jdbcPreparedStatementCreation.end();
+			if ( jdbcPreparedStatementCreation.shouldCommit() ) {
+				jdbcPreparedStatementCreation.executionTime = getExecutionTime( jdbcPreparedStatementCreation.startedAt );
+				jdbcPreparedStatementCreation.sql = preparedStatementSql;
+				jdbcPreparedStatementCreation.commit();
+			}
+		}
+	}
+
+	public static JdbcPreparedStatementExecutionEvent beginJdbcPreparedStatementExecutionEvent() {
+		final JdbcPreparedStatementExecutionEvent jdbcPreparedStatementExecutionEvent = new JdbcPreparedStatementExecutionEvent();
+		if ( jdbcPreparedStatementExecutionEvent.isEnabled() ) {
+			jdbcPreparedStatementExecutionEvent.begin();
+			jdbcPreparedStatementExecutionEvent.startedAt = System.nanoTime();
+		}
+		return jdbcPreparedStatementExecutionEvent;
+	}
+
+	public static void completeJdbcPreparedStatementExecutionEvent(
+			JdbcPreparedStatementExecutionEvent jdbcPreparedStatementExecutionEvent,
+			String preparedStatementSql) {
+		if ( jdbcPreparedStatementExecutionEvent.isEnabled() ) {
+			jdbcPreparedStatementExecutionEvent.end();
+			if ( jdbcPreparedStatementExecutionEvent.shouldCommit() ) {
+				jdbcPreparedStatementExecutionEvent.executionTime = getExecutionTime(
+						jdbcPreparedStatementExecutionEvent.startedAt );
+				jdbcPreparedStatementExecutionEvent.sql = preparedStatementSql;
+				jdbcPreparedStatementExecutionEvent.commit();
+			}
+		}
+	}
+
+	public static void completeJdbcPreparedStatementExecutionEvent(
+			JdbcPreparedStatementExecutionEvent jdbcPreparedStatementExecutionEvent,
+			PreparedStatement statement) {
+		if ( jdbcPreparedStatementExecutionEvent.isEnabled() ) {
+			jdbcPreparedStatementExecutionEvent.end();
+			if ( jdbcPreparedStatementExecutionEvent.shouldCommit() ) {
+				jdbcPreparedStatementExecutionEvent.executionTime = getExecutionTime(
+						jdbcPreparedStatementExecutionEvent.startedAt );
+				jdbcPreparedStatementExecutionEvent.sql = statement.toString();
```

Comment:
This is not safe i.e. `toString` is not guaranteed to return the SQL string. Would be good if you could fix the callsites to use the variant that passes the SQL string directly.

Reply:
this method is called where we do not have access to the SQL String, for all the other cases there is the
```
completeJdbcPreparedStatementExecutionEvent(
			JdbcPreparedStatementExecutionEvent jdbcPreparedStatementExecutionEvent,
			String preparedStatementSql) {
```
method.

Reply:
@beikov see `org.hibernate.engine.jdbc.internal.ResultSetReturnImpl`

- `extract(PreparedStatement statement)`
-  `extract(CallableStatement callableStatement)`
-  `execute(PreparedStatement statement)`
-  `executeUpdate(PreparedStatement statement)`

Reply:
they are all deprecated and it seems we are not using them anymore